ePBFT/src/pbft/libbyz/test/test_ledger_replay.cpp

```diff
@@ -145,70 +150,120 @@ TEST_CASE("Test Ledger Replay")
     replica->big_reqs()->clear();
   }
 
-  INFO("Read the ledger file and replay it out of order and in order");
+  INFO("Read the ledger entries and replay them out of order and in order");
   {
-    auto replay_ledger_io =
-      std::make_unique<asynchost::Ledger>(replay_ledger, wf);
-
-    auto append_ledger_entry_cb =
-      [](const uint8_t* data, size_t size, void* ctx) {
-        auto ledger = static_cast<asynchost::Ledger*>(ctx);
-        ledger->write_entry(data, size);
-      };
-
-    replica->register_append_ledger_entry_cb(
-      append_ledger_entry_cb, replay_ledger_io.get());
-
-    auto initial_ledger_io =
-      std::make_unique<asynchost::Ledger>(initial_ledger, wf);
+    std::vector<uint8_t> initial_ledger;
```

Comment:
I find the interface for the `initial_ledger` a little weird. Is the intent is to have an in-memory ledger? In that case, should we have a common abstract ledger interface and implement an in-memory subclass? I might be reading this wrong though.

Reply:
You are reading it correctly, `initial_ledger` is used to simulate what the class `Ledger` in `ledger.h` does. I had split that class up in the previous PR with intentions of possibly abstracting it, as we had discussed. I can move in this direction of others agree to this.

Reply:
You are reading it correctly, `initial_ledger` is used to simulate what the class `Ledger` in `ledger.h` does. I had split that class up in the previous PR with intentions of possibly abstracting it, as we had discussed. I can move in this direction of others agree to this.

Reply:
I think this is a nice way to have an in-memory equivalent of what we otherwise have in the file, without complicating the code that goes into CCF. There's no use case for an in-memory implementation of the ledger, nor should the enclave code be able to tell where the entries are stored. I prefer to have this in test fixtures for that reason.